src/Test/L0/Worker/TaskManagerL0.cs

```diff
@@ -131,5 +132,153 @@ private Stream GetZipStream()
                 }
             }
         }
+
+        //Test the cancellation flow: interrupt download task via HostContext cancellation token.
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
+        public async void TestCancellationOfEnsureTasksExist()
+        {
+            //Arrange
+            using (var hc = new TestHostContext(this))
+            {
+                hc.SetSingleton<IJobServer>(_jobServer.Object);
+                hc.SetSingleton<IConfigurationStore>(_configurationStore.Object);
+                var taskManager = new TaskManager();
+                taskManager.Initialize(hc);
+
+                _configurationStore.Setup(x => x.GetSettings()).Returns(
+                 new AgentSettings
+                 {
+                     WorkFolder = WorkFolderName
+                 });
+
+                string workingFolder = IOUtil.GetWorkPath(hc);
+                if (Directory.Exists(workingFolder))
+                {
+                    Directory.Delete(IOUtil.GetWorkPath(hc), true);
+                }
+                
+                var bingTask = new TaskInstance()
+                {
+                    Enabled = true,
+                    Name = "Bing",
+                    Version = "0.1.2",
+                    Id = Guid.NewGuid()
+                };
+                var pingTask = new TaskInstance()
+                {
+                    Enabled = true,
+                    Name = "Ping",
+                    Version = "0.1.1",
+                    Id = Guid.NewGuid()
+                };
+
+                var bingVersion = new TaskVersion(bingTask.Version);
+                var pingVersion = new TaskVersion(pingTask.Version);
+
+                _jobServer
+                    .Setup(x => x.GetTaskContentZipAsync(It.IsAny<Guid>(), It.IsAny<TaskVersion>(), hc.CancellationToken))
+                    .Returns((Guid taskId, TaskVersion taskVersion, CancellationToken token) =>
+                    {
+                        hc.Cancel();
+                        return Task.FromResult<Stream>(GetZipStream());
+                    });
+
+                var tasks = new List<TaskInstance>(new TaskInstance[] { bingTask, pingTask });
+
+                //Act
+                //should initiate a download with a mocked IJobServer, which sets a cancelation token and
+                //ensureExistTask task is expected to be in cancelled state
+                Task ensureExistTask = taskManager.EnsureTasksExist(tasks);
+                Task[] taskToWait = { ensureExistTask, Task.Delay(2000) };
```

Comment:
Why Task.Delay(2000)? Can't you just await the task and assert it throws an operation canceled exception? It appears the Task.Delay is extra and not required at all.